src/memo/memo.mli

```diff
@@ -29,19 +44,59 @@ end
     are forgotten, pending computations are cancelled. *)
 val reset : unit -> unit
 
-module type S = Memo_intf.S with type stack_frame := Stack_frame.t
 module type Input = Memo_intf.Input
-module type Output = Memo_intf.Output
-module type Decoder = Memo_intf.Decoder
+module type Data = Memo_intf.Data
 
-module Make(Input : Input)(Decoder : Decoder with type t := Input.t)
-  : S with type input := Input.t
+module Function : sig
+  type ('a, 'b, 'f) t =
+    | Sync : ('a -> 'b) -> ('a, 'b, ('a -> 'b)) t
+    | Async : ('a -> 'b Fiber.t) -> ('a, 'b, ('a -> 'b Fiber.t)) t
+end
+
+module Function_type : sig
+  type ('a, 'b, 'f) t =
+    | Sync : ('a, 'b, ('a -> 'b)) t
+    | Async : ('a, 'b, ('a -> 'b Fiber.t)) t
+end
+
+module Output : sig
+  type 'o t =
+    | Simple of (module Memo_intf.Sexpable with type t = 'o)
+    | Allow_cutoff of (module Data with type t = 'o)
+end
+
+module Visibility : sig
+  type 'i t =
+    | Hidden
+    | Public of 'i Dune_lang.Decoder.t
+end
 
-(** Same as [Make] except that registered functions won't be available
-    through [dune compute]. In particular, it is not necessary to
-    provide a decoder for the input. *)
-module Make_hidden(Input : Input)
-  : S with type input := Input.t
+val create
+  :  string
+  -> doc:string
+  -> input:(module Data with type t = 'i)
+  -> visibility:'i Visibility.t
+  -> output:('o Output.t)
+  -> ('i, 'o, 'f) Function_type.t
+  -> 'f option
+  -> ('i, 'o, 'f) t
+
+(** Set the implementation of a memoized function created with
+    [fcreate] *)
+val set_impl : (_, _, 'f) t -> 'f -> unit
+
+(** Check whether we already have a value for the given call *)
+val peek : ('i, 'o, _) t -> 'i -> 'o option
+val peek_exn : ('i, 'o, _) t -> 'i -> 'o
+
+(** Execute a memoized function *)
+val exec : (_, _, 'f) t -> 'f
+
+(** After running a memoization function with a given name and
+    input, it is possible to query which dependencies that function
+    used during execution by calling [get_deps] with the name and
+    input used during execution. *)
+val get_deps : ('i, _, _) t -> 'i -> (string * Sexp.t) list option
```

Comment:
The return type here is a bit confusing I think. `None` is used to signal that we haven't executed the computation yet, is that correct? Perhaps there should be an explicit type to handle that. Otherwise, it's not clear what is the difference between `None`, `Some []`, etc.

Reply:
Yes, that's correct. I added a doc explicitly saying that.
To me "option" doesn't seem too bad to me in this case. After all, when you're using this function you're already doing something weird, this is not the main mechanism of how people should use memo. 
Also, it was "option" before my change.

I'm not changing it for now, but I'm happy to do it if the above didn't change your mind. Do you prefer polymorphic variant or nominal one?

Reply:
Okay, this change is indeed unrelated. I still think it's a good idea to use a descriptive type, but there's no need to require it for this PR.